lib/migrate_user_config.py

```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env python3
+"""
+This will run in preinst when upgrading from any version prior to 27.13.3.
+It removes the "ua_config" sub-document from uaclient.conf and moves it to it's own json file in /var/lib.
+The fields under "ua_config" are intended to be set/unset by the user, so we don't want them coverred as a "conffile".
+
+This script does a best effort to edit uaclient.conf in such a way that it will result in no conffile prompts. It writes
+back a uaclient.conf that will match the default conf as long as no defaults have changed.
+"""
+
+"""
+for dev reference: TODO delete
+old uaclient.conf:
+# Ubuntu Pro Client config file.
+# If you modify this file, run "pro refresh config" to ensure changes are
+# picked up by Ubuntu Pro Client.
+
+contract_url: https://contracts.canonical.com
+data_dir: /var/lib/ubuntu-advantage
+log_file: /var/log/ubuntu-advantage.log
+log_level: debug
+security_url: https://ubuntu.com/security
+timer_log_file: /var/log/ubuntu-advantage-timer.log
+daemon_log_file: /var/log/ubuntu-advantage-daemon.log
+ua_config:
+  apt_http_proxy: null
+  apt_https_proxy: null
+  http_proxy: null
+  https_proxy: null
+  update_messaging_timer: 21600
+  metering_timer: 14400
+
+new uaclient.conf:
+contract_url: https://contracts.canonical.com
+security_url: https://ubuntu.com/security
+log_level: debug
+"""
+
+import json
+import logging
+
+import yaml
+
+from uaclient import defaults
+from uaclient.daemon import setup_logging
+
+
+def main():
+    # TODO probs try/except anything here
+    setup_logging(logging.WARN, logging.DEBUG)
+    uaclient_conf_path = "/etc/ubuntu-advantage/uaclient.conf"
+    with open(uaclient_conf_path, "r") as uaclient_conf_file:
+        old_uaclient_conf = yaml.safe_load(uaclient_conf_file)
+
+    new_uaclient_conf = {
+        "contract_url": old_uaclient_conf.get(
+            "contract_url", defaults.BASE_CONTRACT_URL
+        ),
+        "security_url": old_uaclient_conf.get(
+            "security_url", defaults.BASE_SECURITY_URL
+        ),
+        "log_level": old_uaclient_conf.get("log_level", "debug"),
```

Comment:
Quick  question, considering we transform the log_level to upper in [config](https://github.com/canonical/ubuntu-advantage-client/blob/12c12d0c19bf55cd5f0f48148a7cdd8b1dfb8a3b/uaclient/config.py#L363), shouldn't we not then default to capitalized version of `debug` as well?

Reply:
Alternatively: because we transform to upper in config, why do we need to capitalize the value in the file?

IMO: it doesn't matter, but the current default is `debug` and lowercase is easier to type :upside_down_face: so I'd prefer to keep it lowercase here.